modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +76,47 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): F[Set[P2PContext]] = for {
+    _ <- twoWayHandshake(context, none).attempt.map(_.left.map { err =>
+      logger.error(err)(s"Joining to P2PContext ${context.show} failed")
+    })
+    maybePeer <- clusterStorage.getPeer(context.id)
+    maybeDiscovered = maybePeer.map {
+      peerDiscovery
+        .discoverFrom(_)
+        .map(_.map(toP2PContext))
+        .handleErrorWith { err =>
+          logger
+            .error(err)(s"Peer discovery from P2PContext ${context.show} failed")
+            .as(Set.empty)
+        }
+    }
+    discovered <- maybeDiscovered.getOrElse(Set.empty[P2PContext].pure)
```

Comment:
This might be cleaner if you use `OptionT`.

Reply:
Implemented the suggested re-factor.